engine/src/ui/canvas/plugin.cpp

```diff
@@ -0,0 +1,118 @@
+#include <glm/gtc/matrix_transform.hpp>
+
+#include <cubos/core/io/window.hpp>
+
+#include <cubos/engine/render/target/plugin.hpp>
+#include <cubos/engine/render/target/target.hpp>
+#include <cubos/engine/transform/child_of.hpp>
+#include <cubos/engine/transform/plugin.hpp>
+#include <cubos/engine/ui/canvas/horizontal_stretch.hpp>
+#include <cubos/engine/ui/canvas/plugin.hpp>
+#include <cubos/engine/ui/canvas/ui_canvas.hpp>
+#include <cubos/engine/ui/canvas/ui_element.hpp>
+#include <cubos/engine/ui/canvas/vertical_stretch.hpp>
+#include <cubos/engine/window/plugin.hpp>
+
+CUBOS_DEFINE_TAG(cubos::engine::elementVPUpdateTag);
+CUBOS_DEFINE_TAG(cubos::engine::elementPropagateTag)
+CUBOS_DEFINE_TAG(cubos::engine::uiBeginTag)
+CUBOS_DEFINE_TAG(cubos::engine::uiDrawTag)
+
+using namespace cubos::engine;
+
+void cubos::engine::canvasPlugin(Cubos& cubos)
+{
+    cubos.component<UIElement>();
+    cubos.component<UICanvas>();
+    cubos.component<HorizontalStretch>();
+    cubos.component<VerticalStretch>();
+
+    cubos.depends(windowPlugin);
+    cubos.depends(renderTargetPlugin);
+    cubos.depends(transformPlugin);
+
+    cubos.tag(elementVPUpdateTag);
+    cubos.tag(elementPropagateTag).after(elementVPUpdateTag);
+    cubos.tag(uiBeginTag).after(elementPropagateTag);
+    cubos.tag(uiDrawTag).after(uiBeginTag).before(windowRenderTag);
+
+    cubos.observer("initialize Canvas").onAdd<UICanvas>().call([](Query<UICanvas&> query) {
+        for (auto [canvas] : query)
+        {
+            canvas.mat = glm::ortho<float>(0, canvas.width, 0, canvas.height, -10000, 100000);
```

Comment:
These arbitrary min/max values scare me a little. Do we need such a big range? And in both directions?

Reply:
Well, yeah, I think we need a big range, at least if we're using the z coordinates method for sorting. If you have any ideas for an alternative way, lmk

Reply:
Do we need it in both directions? Also, we have to be careful with precision loss, it may lead to z-fighting. Maybe its safer to keep the limits around ~1000.

Reply:
In my experience, having both directions helps a lot. Imagine after you already have your scene with ~600 ui elements you find out you need another one to show below everything. You can either mode every single ui elements one more layer down, or you can just go into the negatives.